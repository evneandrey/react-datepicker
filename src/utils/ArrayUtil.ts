export const chunk = (arr: any[], n: number) => {
  const result = [];
  let i = 0;
  while (i < arr.length / n) {
    result.push(arr.slice(i * n, i * n + n));
    i += 1;
  }

  return result;
};

export const range = (n1: number, n2?: number) => {
  const result = [];
  let first = !n2 ? 0 : n1;
  let last = n2;

  if (!last) {
    last = n1;
  }

  while (first < last) {
    result.push(first);
    first += 1;
  }
  return result;
};

export const repeat = (el: any, n: number) => {
  return range(n).map(() => el);
};
