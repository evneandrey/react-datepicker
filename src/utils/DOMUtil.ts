import { IDatePicker } from '../common/@types';
const convertPx = (value: number) => `${value}px`;
/**
 * Getting Div position as far as distance
 * @param node
 * @param direction
 * @param distance
 */
export const getDivPosition = (
  node: HTMLDivElement | null,
  direction: IDatePicker.PickerDirection = IDatePicker.PickerDirection.BOTTOM,
  height: number,
  distance: number = 5
): IDatePicker.Position => {
  if (!node) return { left: '', top: '', bottom: '' };

  let top = 0;
  let left = 0;

  switch (direction) {
    case IDatePicker.PickerDirection.BOTTOM:
      top = node.offsetTop + node.offsetHeight + distance;
      left = node.offsetLeft;
      break;
    case IDatePicker.PickerDirection.TOP:
      top = node.offsetTop - height - distance;
      left = node.offsetLeft;
      break;
  }

  return {
    top: convertPx(top),
    left: convertPx(left),
  };
};

export const getDomHeight = (node: HTMLDivElement | null): number => {
  return node ? node.clientHeight : 0;
};
