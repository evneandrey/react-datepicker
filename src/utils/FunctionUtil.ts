export const ifExistCall = (func?: (...args: any[]) => void, ...args: any[]) =>
  func && func(...args);
