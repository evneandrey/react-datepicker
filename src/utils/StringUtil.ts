export const lpad = (val: string, length: number, char: string = '0') =>
  val.length < length ? char.repeat(length - val.length) + val : val;
