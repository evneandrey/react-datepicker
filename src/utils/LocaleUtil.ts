import * as dayjs from 'dayjs';
import { IDatePicker } from '../common/@types';
import { range } from '../utils/ArrayUtil';
import * as localeData from 'dayjs/plugin/localeData';
import * as localizedFormat from 'dayjs/plugin/localizedFormat';

dayjs.extend(localeData);
dayjs.extend(localizedFormat);

export const getMonthShort = (locale: IDatePicker.Locale) => {
  dayjs.locale(locale);
  return range(0, 12).map(v =>
    dayjs()
      .localeData()
      .monthsShort(dayjs().month(v))
  );
};

export const getWeekDays = (locale: IDatePicker.Locale) => {
  dayjs.locale(locale);
  return range(7).map(v =>
    dayjs()
      .localeData()
      .weekdaysShort(dayjs().day(v))
  );
};

export const getToday = (locale: IDatePicker.Locale) => {
  return dayjs()
    .locale(locale)
    .format('LL');
};
