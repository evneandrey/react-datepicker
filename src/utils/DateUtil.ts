import { chunk, repeat, range } from './ArrayUtil';
import { IDatePicker } from '../common/@types';
import * as dayjs from 'dayjs';
import { getMonthShort } from './LocaleUtil';

export const getDayMatrix = (year: number, month: number): string[][] => {
  const date = dayjs()
    .year(year)
    .month(month);

  const startOfMonth = date.startOf('month').date();
  const endOfMonth = date.endOf('month').date();

  const startDay = date.startOf('month').day();
  const remain = (startDay + endOfMonth) % 7;

  return chunk(
    [
      ...repeat(' ', startDay),
      ...range(startOfMonth, endOfMonth + 1).map(v => `${v}`),
      ...(7 - remain === 7 ? [] : repeat(' ', 7 - remain)),
    ],
    7
  );
};

export const getMonthMatrix = (locale: IDatePicker.Locale) => {
  return chunk(getMonthShort(locale), 3);
};

export const getYearMatrix = (year: number): string[][] => {
  return chunk(range(year - 4, year + 5).map(v => `${v}`), 3);
};

export const isDayEqual = (day1?: dayjs.Dayjs, day2?: dayjs.Dayjs) => {
  if (!day1 || !day2) return false;
  return dayjs(day1).isSame(day2, 'date');
};

export const isDayAfter = (day1: dayjs.Dayjs, day2: dayjs.Dayjs) => {
  return dayjs(day1).isAfter(day2, 'date');
};

export const isDayBefore = (day1: dayjs.Dayjs, day2: dayjs.Dayjs) => {
  return dayjs(day1).isBefore(day2, 'date');
};

export const isDayRange = (date: dayjs.Dayjs, start?: dayjs.Dayjs, end?: dayjs.Dayjs) => {
  if (!start || !end) return false;

  return isDayAfter(date, start) && isDayBefore(date, end);
};

export const formatDate = (date: dayjs.Dayjs | undefined, format: string) => {
  if (date === undefined) return '';
  return dayjs(date).format(format);
};
