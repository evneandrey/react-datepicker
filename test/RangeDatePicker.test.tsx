import * as React from 'react';
import * as sinon from 'sinon';
import * as dayjs from 'dayjs';
import { shallow, mount, ReactWrapper } from 'enzyme';
import RangeDatePicker, { Props, State } from '../src/components/RangeDatePicker';
import { mountInputSimulateChange } from './utils/TestingUtil';

const START_INPUT_CLASS = '.range-picker-input__start .picker-input__text';
const END_INPUT_CLASS = '.range-picker-input__end .picker-input__text';
const INPUT_FORMAT = 'YYYY-MM-DD';

describe('<RangeDatePicker/>', () => {
  const defaultProps = {
    initialDate: dayjs(new Date(2018, 4, 1)),
  };

  const pickerShow = (component: ReactWrapper) => {
    component.find('Picker').setState({
      show: true,
    });
  };

  describe('shallow test', () => {
    it('should render without crash', () => {
      const shallowComponent = shallow(<RangeDatePicker />);
      expect(shallowComponent).toBeTruthy();
      expect(shallowComponent).toMatchSnapshot();
    });
  });

  describe('handleDateChange', () => {
    let mountComponent: ReactWrapper<Props, State>;
    const dayClick = (calendarAt: number) => (dayAt: number) =>
      mountComponent
        .find('.calendar__item')
        .at(calendarAt)
        .find('td')
        .at(dayAt)
        .simulate('click');

    it('should first select then state start change', () => {
      mountComponent = mount(<RangeDatePicker {...defaultProps} />);
      mountComponent.find(START_INPUT_CLASS).simulate('click');
      dayClick(0)(2);
      const start = mountComponent.state('start');
      expect(start).not.toBeUndefined();
      if (start) {
        expect(dayjs(start).format('YYYYMMDD')).toEqual('20180501');
      }
    });

    it('should state start be set & select start before then state start change', () => {
      mountComponent = mount(<RangeDatePicker {...defaultProps} />);
      mountComponent.find(START_INPUT_CLASS).simulate('click');
      mountComponent.setState({
        start: dayjs(new Date(2018, 4, 5)),
      });
      dayClick(0)(2);
      const start = mountComponent.state('start');
      expect(start).not.toBeUndefined();
      if (start) {
        expect(dayjs(start).format('YYYYMMDD')).toEqual('20180501');
      }
    });

    it('should state start be set & select start after then state end change', () => {
      mountComponent = mount(<RangeDatePicker {...defaultProps} />);
      mountComponent.find(START_INPUT_CLASS).simulate('click');
      mountComponent.setState({
        start: dayjs(new Date(2018, 4, 2)),
      });
      dayClick(0)(5);
      const end = mountComponent.state('end');
      expect(end).not.toBeUndefined();
      if (end) {
        expect(dayjs(end).format('YYYYMMDD')).toEqual('20180504');
      }
    });

    it('should state start, end be set then state start change', () => {
      mountComponent = mount(<RangeDatePicker {...defaultProps} />);
      mountComponent.find(START_INPUT_CLASS).simulate('click');
      mountComponent.setState({
        start: dayjs(new Date(2019, 4, 2)),
        end: dayjs(new Date(2018, 4, 9)),
      });
      dayClick(0)(5);
      const start = mountComponent.state('start');
      const end = mountComponent.state('end');
      expect(start).not.toBeUndefined();
      expect(end).toBeUndefined();
      if (start) {
        expect(dayjs(start).format('YYYYMMDD')).toEqual('20180504');
      }
    });

    it('should prop onChange fire correctly', () => {
      const onChange = sinon.spy();
      mountComponent = mount(<RangeDatePicker onChange={onChange} {...defaultProps} />);
      mountComponent.find(START_INPUT_CLASS).simulate('click');
      dayClick(0)(5);
      expect(onChange).toHaveProperty('callCount', 1);
    });
  });

  describe('handleInputChange', () => {
    let mountComponent: ReactWrapper<Props, State>;

    beforeEach(() => {
      mountComponent = mount(<RangeDatePicker dateFormat={INPUT_FORMAT} {...defaultProps} />);
    });

    it('should start & end input change set state value', () => {
      const startInput = mountComponent.find(START_INPUT_CLASS);
      const endInput = mountComponent.find(END_INPUT_CLASS);
      const testValue = 'test';
      mountInputSimulateChange(startInput, testValue);
      expect(mountComponent.state('startValue')).toEqual(testValue);

      mountInputSimulateChange(endInput, testValue);
      expect(mountComponent.state('endValue')).toEqual(testValue);
    });
  });

  describe('handleInputBlur', () => {
    let mountComponent: ReactWrapper<Props, State>;
    const start = dayjs(new Date(2018, 4, 1));
    const end = dayjs(new Date(2018, 4, 11));

    beforeEach(() => {
      mountComponent = mount(<RangeDatePicker dateFormat={INPUT_FORMAT} {...defaultProps} />);
    });

    it('should state start & end undefined return empty value', () => {
      mountComponent.setState({
        ...mountComponent.state,
        start: undefined,
        end: undefined,
      });

      mountComponent.find(START_INPUT_CLASS).simulate('blur');
      expect(mountComponent.state('startValue')).toEqual('');

      mountComponent.find(END_INPUT_CLASS).simulate('blur');
      expect(mountComponent.state('endValue')).toEqual('');
    });

    it('should start input invalid value recover original date', () => {
      mountComponent.setState({
        ...mountComponent.state,
        start,
        end,
        startValue: '20183343j',
        endValue: dayjs(end).format(INPUT_FORMAT),
      });

      mountComponent.find(START_INPUT_CLASS).simulate('blur');
      expect(mountComponent.state('startValue')).toEqual('2018-05-01');
    });

    it('should start input correct value set state date', () => {
      const changedValue = '2018-04-23';
      mountComponent.setState({
        ...mountComponent.state,
        start,
        end,
        startValue: changedValue,
        endValue: dayjs(end).format(INPUT_FORMAT),
      });

      mountComponent.find(START_INPUT_CLASS).simulate('blur');
      const startState = mountComponent.state('start');
      expect(startState).not.toBeNull();
      if (startState) {
        expect(dayjs(startState).format(INPUT_FORMAT)).toEqual(changedValue);
      }
    });

    it('should end input invalid value recover original date', () => {
      mountComponent.setState({
        ...mountComponent.state,
        start,
        end,
        startValue: dayjs(start).format(INPUT_FORMAT),
        endValue: '201804244',
      });

      mountComponent.find(END_INPUT_CLASS).simulate('blur');
      expect(mountComponent.state('endValue')).toEqual('2018-05-11');
    });

    it('should end input correct value set state date', () => {
      const changedValue = '2018-05-23';
      mountComponent.setState({
        ...mountComponent.state,
        start,
        end,
        startValue: dayjs(start).format(INPUT_FORMAT),
        endValue: changedValue,
      });

      mountComponent.find(END_INPUT_CLASS).simulate('blur');
      const endState = mountComponent.state('end');
      expect(endState).not.toBeNull();
      if (endState) {
        expect(dayjs(endState).format(INPUT_FORMAT)).toEqual(changedValue);
      }
    });

    it('should startDate < endDate swap date', () => {
      const endInput = mountComponent.find(END_INPUT_CLASS);
      const testValue = '2018-04-23';
      // 2018-04-01
      const startValue = dayjs(start).format(INPUT_FORMAT);
      mountComponent.setState({
        ...mountComponent.state,
        start,
        end,
        startValue,
        endValue: testValue,
      });

      endInput.simulate('blur');

      const startState = mountComponent.state('start');
      const endState = mountComponent.state('end');

      expect(startState).not.toBeNull();
      expect(endState).not.toBeNull();

      if (startState && endState) {
        expect(dayjs(startState).format(INPUT_FORMAT)).toEqual(testValue);
        expect(dayjs(endState).format(INPUT_FORMAT)).toEqual(startValue);
      }
    });

    it('should startDate > endDate swap date', () => {
      const startInput = mountComponent.find(START_INPUT_CLASS);
      const testValue = '2018-06-23';
      // 2018-04-01
      const endValue = dayjs(end).format(INPUT_FORMAT);
      mountComponent.setState({
        ...mountComponent.state,
        start,
        end,
        endValue,
        startValue: testValue,
      });

      startInput.simulate('blur');

      const startState = mountComponent.state('start');
      const endState = mountComponent.state('end');

      expect(startState).not.toBeNull();
      expect(endState).not.toBeNull();

      if (startState && endState) {
        expect(dayjs(startState).format(INPUT_FORMAT)).toEqual(endValue);
        expect(dayjs(endState).format(INPUT_FORMAT)).toEqual(testValue);
      }
    });
  });

  describe('handleInputClear', () => {
    let mountComponent: ReactWrapper;
    const rangeInputClass = (value: string) => `.range-picker-input__${value}`;
    const start = dayjs(new Date(2018, 4, 1));
    const end = dayjs(new Date(2018, 4, 11));

    beforeEach(() => {
      mountComponent = mount(<RangeDatePicker clear {...defaultProps} />);
    });

    it('should start input clear click start & startValue init correctly', () => {
      mountComponent.setState({
        ...mountComponent.state,
        start,
        startValue: dayjs(start).format(INPUT_FORMAT),
      });
      mountComponent
        .find(rangeInputClass('start'))
        .find('.icon-clear')
        .first()
        .simulate('click');

      const startState = mountComponent.state('start');
      const startValue = mountComponent.state('startValue');

      if (startState) {
        expect(startState).toBeUndefined();
        expect(startValue).toEqual('');
      }
    });

    it('should end input clear click end & endValue init correctly', () => {
      mountComponent.setState({
        ...mountComponent.state,
        end,
        endValue: dayjs(start).format(INPUT_FORMAT),
      });
      mountComponent
        .find(rangeInputClass('end'))
        .find('.icon-clear')
        .first()
        .simulate('click');

      const endState = mountComponent.state('end');
      const endValue = mountComponent.state('endValue');

      if (endState) {
        expect(endState).toBeUndefined();
        expect(endValue).toEqual('');
      }
    });
  });

  describe('handleMouseOver', () => {
    let mountComponent: ReactWrapper<Props, State>;

    beforeEach(() => {
      mountComponent = mount(<RangeDatePicker {...defaultProps} />);
    });

    it('does mouseover event set state hoverDate', () => {
      pickerShow(mountComponent);
      mountComponent
        .find('td')
        .at(5)
        .simulate('mouseover');
      const hoverDate = mountComponent.state('hoverDate');
      expect(hoverDate).not.toBeUndefined();
      if (hoverDate) {
        expect(dayjs(hoverDate).format('YYYYMMDD')).toEqual('20180504');
      }
    });

    it('should mouseover range hover correctly', () => {
      pickerShow(mountComponent);
      mountComponent.setState({
        ...mountComponent.state,
        start: dayjs(new Date(2018, 4, 1)),
      });

      mountComponent
        .find('td')
        .at(5)
        .simulate('mouseover');

      expect(
        mountComponent
          .find('td')
          .at(4)
          .hasClass('calendar__day--range')
      ).toBeTruthy();
    });
  });

  describe('Calendar Wrapper', () => {
    let mountComponent: ReactWrapper;

    beforeEach(() => {
      const wrapper = (calendar: JSX.Element) => {
        const container = <div className="wrapper-test">{calendar}</div>;
        return container;
      };
      mountComponent = mount(<RangeDatePicker {...defaultProps} wrapper={wrapper} />);
    });

    it('should calendar wrapper render without crash', () => {
      pickerShow(mountComponent);
      expect(mountComponent.find('.wrapper-test')).toHaveLength(1);
    });
  });
});
