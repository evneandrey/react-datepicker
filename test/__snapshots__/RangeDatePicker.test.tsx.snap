// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RangeDatePicker/> shallow test should render without crash 1`] = `
<Picker
  portal={false}
  renderContents={[Function]}
  renderTrigger={[Function]}
/>
`;
